x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java

```diff
@@ -835,6 +845,19 @@ void start() {
                         EsExecutors.rethrowErrors(contextHolder.unwrap(runnable));
                     }
                 }
+
+                // if shutdown with tasks pending notify the handlers
+                if (queue.isEmpty() == false) {
+                    List<Runnable> notExecuted = new ArrayList<>();
+                    queue.drainTo(notExecuted);
+
+                    for (Runnable runnable : notExecuted) {
+                        if (runnable instanceof AbstractRunnable) {
+                            ((AbstractRunnable) runnable).onRejection(
+                                    new EsRejectedExecutionException("unable to process as autodetect worker service has shutdown", true));
+                        }
+                    }
+                }
```

Comment:
I think there's still a race condition, although probably less likely than it was before.  Consider this scenario:

1. Thread 1 is running `start()`, and is preempted on line 841.
2. Thread 2 is calling `execute()`, and is preempted on line 828 while `running == true`.
3. Thread 3 calls `shutdown()` and sets `running = false`.
4. Thread 1 goes back on CPU and progresses to line 860, i.e. all the way through the `// if shutdown with tasks pending notify the handlers` block.
5. Thread 2 goes back on CPU and adds its `command` to the `queue`.

Step 5 happens after the queue is drained, so the problem this PR is trying to fix will still occur - its `command` is never run nor rejected.

One way to fix this race would be to make the whole of `execute()` `synchronized`, plus the queue draining part of `start()`, i.e. lines 849-860.

Reply:
Good point, there is a lesson here about rolling your own instead of using a proven executor. 

I looked at using a fixed threadpool of size 1 from `EsExecutors.newFixed` but that method wants ThreadFactory whereas here we use a single thread from the ml autodetect pool. My preference is to use one of the standard Executors but I can't see how to make that work